lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp

```diff
@@ -232,20 +232,62 @@ sharedToLinearLayoutAMDRotating(ArrayRef<int64_t> shape,
 
 } // namespace
 
+static LinearLayout reshapeLayout(MLIRContext *ctx, LinearLayout layout,
+                                  ArrayRef<int64_t> shape) {
+  int rank = shape.size();
+  auto srcOutDims = to_vector(layout.getOutDimNames());
+  std::reverse(srcOutDims.begin(), srcOutDims.end());
+  auto newOutDims = standardOutDimPairs(ctx, shape);
+  std::reverse(newOutDims.begin(), newOutDims.end());
+  return layout.transposeOuts(srcOutDims)
+      .reshapeOuts(newOutDims)
+      .transposeOuts(standardOutDimNames(ctx, rank));
+}
+
+static LinearLayout transposeLinearLayout(LinearLayout layout,
+                                          ArrayRef<int> order) {
+  // Transpose the tile layout.
+  auto namedBases = layout.getBases();
+  // move the most outer dimensions to the inner most position.
+
+  for (auto &bases : llvm::make_second_range(namedBases)) {
+    for (auto &b : bases) {
+      std::vector<int32_t> newB;
+      for (auto i : order) {
+        newB.push_back(b[i]);
+      }
+      b = std::move(newB);
+    }
+  }
+  return LinearLayout(std::move(namedBases),
+                      to_vector(layout.getOutDimNames()));
+}
+
 LinearLayout sharedToLinearLayoutLeadingOffset(ArrayRef<int64_t> shape,
                                                NVMMASharedEncodingAttr shared,
                                                bool disableSwizzle) {
   MLIRContext *ctx = shared.getContext();
-
-  auto shapePerCTA = getShapePerCTA(shared, shape);
-
   int rank = shape.size();
+  auto shapePerCTA = getShapePerCTA(shared, shape);
   if (rank == 1) {
     // TODO: Not sure if this is correct.
     return combineCtaCgaWithShape(
         LinearLayout::identity1D(shapePerCTA[0], S("offset"), S("dim0")),
         shared.getCTALayout(), shape);
   }
+  // Construct bases for a the layout's 2-dimensional tile.
+  assert(rank >= 2);
+  int batchDims = rank - 2;
+
+  // Collapse all the outer dim into one. We will then create a layout for this
```

Comment:
@peterbell10 @lezcano this is a bit of a mess but I have been breaking my head trying to make this work. 
I believe we need to also redo that with the patch Peter is working on to support higher dim than 256 so I didn't want to spend more time than needed on something that will change.
I need to relax cases where rank > 2 and the second most inner dim is less than 8, the swizzle still works as we can swizzle on more than just 2 dimensions. To make it work I collapse the higher dim but once we want to support dim size > 256 we will have to handle all the dimension in this logic.
I'm happy to help clean up the mess later.

Reply:
happy to tidy all this up once Peter has reworked his part